### PR TITLE
improve debug symbol discovery

### LINKED ISSUE #265: libc debug information is not taken from Ubuntu libc6-dbg
**Describe the bug**
The symbols from system libraries such as libc are not properly imported into the profiling results.

**To Reproduce**
1. Do `sudo apt install libc6-dbg` and try to profile a code, which calls a libc function
2. Despite the debug information is available, the libc symbols will be shown "??" in hotspot (see screenshot)

**Expected behavior**
The actual libc symbol name is expected instead of ??

**Version Info (please complete the following information):**
 - Linux Kernel version: 5.4.0
 - perf version: 5.4.60
 - hotspot version: AppImage

**Additional context**

Screenshot from 2020-10-21 12-01-02

### PR DESCRIPTION
**Pull request by lievenhey**
If libunwind fails to find the debug symbols, the perfparser tries to
find it by checking first in /usr/lib/ than in .debug and than in
/usr/lib/debug. This causes some problems if the debug file has the same
name as the library itself. In this case the algorithm stops at
/usr/lib/ and returns the wrong file.
To prevent this I reorder the algorithm to first check in .debug than in
/usr/lib/debug and than in /usr/lib/

I also changed how the algorithm searches inside /usr/lib/debug. Some
distros (e.g. Ubuntu) uses shorted path names so the debug file for
/usr/lib/x86_64-linux-gnu/libc-2.31.so is located in
/usr/lib/debug/lib/x86_64-linux-gnu/libc-2.31.so. To find the file the
algorithm uses combines /usr/lib/debug/ with the path to the original
file. If the files does not exists it removes the first directory from
the path and tries again. (in case of the libc example the new path
wouls be /lib/x86_64-linux-gnu/libc-2.31.so)

fixes KDAB/hotspot#265

### CODE REVIEW

File: app/perfsymboltable.cpp [hunk 1]
```diff
@@ -497,9 +487,36 @@ static QFileInfo findDebugInfoFile(const QString &root, const QString &file,
         return debugLinkFile;
 
     // try again in /usr/lib/debug folder
-    debugLinkFile.setFile(dir, QLatin1String("usr") + QDir::separator() + QLatin1String("lib")
-                                + QDir::separator() + QLatin1String("debug") + QDir::separator() + folder
-                                + QDir::separator() + debugLinkString);
+    auto path = folder;
+    while (path.contains(QDir::separator())) {
```
Comment (milianw) [inline app/perfsymboltable.cpp]: what is this loop trying to do?  can you add a comment? it looks like this tries to find

/foo/bar/file.so
/bar/file.so
/file.so

each prefixed with the `usr/lib/debug` path. right? why is this only done for this folder, and not the `.debug` folder above?

and can you please give a concrete example in the comment - i.e. something like "on ubuntu x.y a debug link is given as X/Y/Z but we need to find usr/lib/debug/Y/Z instead?

I also believe that this loop does not work correctly if `folder` contains no separator, as it would then never try to append that file stem to the search path

Comment (milianw) [reply app/perfsymboltable.cpp]: please also fix the commit message: we don't use libunwind, you probably mean elfutils or libdwfl

Comment (milianw) [reply app/perfsymboltable.cpp]: > Some distros (e.g. Ubuntu) uses shorted path names so the debug file for /usr/lib/x86_64-linux-gnu/libc-2.31.so is located in /usr/lib/debug/lib/x86_64-linux-gnu/libc-2.31.so.

Isn't that rather a situation where the long path is a symlink to the short path, and it uses canonical paths in the debug dir?

File: app/perfsymboltable.cpp [hunk 2]
```diff
@@ -475,14 +474,14 @@ static QFileInfo findDebugInfoFile(const QString &root, const QString &file,
 
     // try in .debug folder
     if (!folder.isEmpty()) {
-        debugLinkFile.setFile(dir, folder + QDir::separator() + QLatin1String(".debug")
-                                    + QDir::separator() + debugLinkString);
+        debugLinkFile.setFile(dir.path() + QDir::separator() + folder + QDir::separator() + QLatin1String(".debug")
```
Comment (milianw) [inline app/perfsymboltable.cpp:478]: why this change?

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
+/****************************************************************************
+**
+** Copyright (C) 2017 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com, author Milian Wolff
+*<milian.wolff@kdab.com>
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp]: make yourself the author

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                 == QFileInfo(tempDir->path() + QStringLiteral("/usr/lib/libqt.so")).absoluteFilePath());
+    }
+
+    void cleanupTestCase()
+    {
+        delete tempDir;
+    }
+
+private:
+    QTemporaryDir* tempDir;
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp]: make it a plain member, not a pointer

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    void testDefaultFile()
+    {
+        auto libqtDebug = findDebugInfoFile(tempDir->path() + QStringLiteral("/"), QStringLiteral("/usr/lib/libqt.so"),
+                                            QStringLiteral("libqt.so"));
+        Q_ASSERT(libqtDebug.absoluteFilePath()
+                 == QFileInfo(tempDir->path() + QStringLiteral("/usr/lib/libqt.so")).absoluteFilePath());
+    }
+
+    void cleanupTestCase()
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp]: remove this then

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+QFileInfo findDebugInfoFile(const QString& root, const QString& file, const QString& debugLinkString);
+
+class TestFindDebugSymbols : public QObject
+{
+    Q_OBJECT
+private slots:
+    void initTestCase()
+    {
+        tempDir = new QTemporaryDir();
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp]: won't be needed anymore (see below)

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            QStringLiteral("/usr/lib/debug/libm.so"),
+            QStringLiteral("/usr/lib/x64/libc.so"),
+            QStringLiteral("/usr/lib/debug/lib/x64/libc.so"),
+            QStringLiteral("/usr/lib/libqt.so"),
+        };
+
+        QDir dir(tempDir->path());
+        for (const auto& file : files) {
+            auto path = QFileInfo(tempDir->path() + file).path();
+            dir.mkpath(path);
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp:52]: QVERIFY()

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            QFile f(tempDir->path() + file);
+            f.open(QIODevice::WriteOnly);
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp:56]: QVERIFY()

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            f.write(file.toUtf8());
+            f.close();
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp]: not needed, will get closed automatically

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        }
+    }
+
+    void testDotDebug()
+    {
+        auto pythonDebug =
+            findDebugInfoFile(tempDir->path() + QStringLiteral("/"), QStringLiteral("/usr/bin/python3.8"),
+                              QStringLiteral("096cdc8214a805dca8d174fe072684b0f21645.debug"));
+        Q_ASSERT(pythonDebug.absoluteFilePath()
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp]: QCOMPARE instead

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+            auto path = QFileInfo(tempDir->path() + file).path();
+            dir.mkpath(path);
+            QFile f(tempDir->path() + file);
+            f.open(QIODevice::WriteOnly);
+            f.write(file.toUtf8());
+            f.close();
+        }
+    }
+
+    void testDotDebug()
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp]: all of these tests should be data driven, cf. QTest::addColumn/addRow and QFETCH

then it should just be a matter of feeding the data to be passed to the findDebugInfoFile call as columns in the various rows. and the test itself just QCOMPARES the result against the expected value you also put into a column value

**Review by milianw — CHANGES_REQUESTED** [review]

File: app/perfsymboltable.cpp (hunk 2 above)
Comment (lievenhey) [reply app/perfsymboltable.cpp:477]: got tricked by an ui bug which let me believe that this code does not work

Comment (lievenhey) [reply app/perfsymboltable.cpp:478]: If I call QFileInfo::setFile with dir is `/path/to/sysroot` and folder is `/usr/lib/` QFileInfo will point to `/usr/lib/...` and not `/path/to/sysroot/usr/lib/...`

File: tests/auto/finddebugsym/tst_finddebugsym.cpp [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+                                        << QStringLiteral("/usr/lib/debug/libm.so");
+        QTest::addRow("/usr/lib/debug") << QStringLiteral("/usr/lib/x64/libc.so") << QStringLiteral("libc.so")
+                                        << QStringLiteral("/usr/lib/debug/lib/x64/libc.so");
+        QTest::addRow("no debug file") << QStringLiteral("/usr/lib/libqt.so") << QStringLiteral("libqt.so")
+                                       << QStringLiteral("/usr/lib/libqt.so");
+    }
+
+    void findDebugSymbols()
+    {
+        QFETCH(QString, file);
```
Comment (milianw) [inline tests/auto/finddebugsym/tst_finddebugsym.cpp:86]: the names for these columns are super confusing, this should be `root, file, debugLinkString` instead

**Review by milianw — CHANGES_REQUESTED** [review]

File: app/perfsymboltable.cpp (hunk 1 above)
Comment (lievenhey) [reply app/perfsymboltable.cpp]: yeah, changed it (thanks for the hint). Now it creates a copy of the path with `/usr/` removed from the file path and checks in a for both paths (with and without `usr`) if the file exists in `/usr/lib/debug`

File: app/perfsymboltable.cpp [hunk 13]
```diff
@@ -508,19 +536,23 @@ int PerfSymbolTable::findDebugInfo(Dwfl_Module *module, const char *moduleName,
...
+    findDebugInfoFile(m_unwind->systemRoot(), path, debugLinkString);
     int ret = dwfl_standard_find_debuginfo(module, nullptr, moduleName, base, file,
                                            debugLink, crc, debugInfoFilename);
     if (ret >= 0 || !debugLink || strlen(debugLink) == 0)
         return ret;
 
     // fall-back, mostly for situations where we loaded a file via it's build-id.
     // search all known paths for the debug link in that case
-    const auto debugLinkString = QFile(debugLink).fileName();
+    // const auto debugLinkString = QFile(debugLink).fileName();
```
Comment (milianw) [inline app/perfsymboltable.cpp]: this needs to be cleaned up

File: app/perfsymboltable.cpp [hunk 14]
```diff
@@ -533,15 +565,15 @@ int PerfSymbolTable::findDebugInfo(Dwfl_Module *module, const char *moduleName,
     if (!debugLinkFile.isFile())
         return ret;
 
-    const auto path = eu_compat_strdup(debugLinkFile.absoluteFilePath().toUtf8().constData());
+    /*const auto path = eu_compat_strdup(debugLinkFile.absoluteFilePath().toUtf8().constData());
```
Comment (milianw) [inline app/perfsymboltable.cpp]: dito

File: app/perfsymboltable.cpp [hunk 15]
```diff
@@ -473,33 +473,61 @@ static QFileInfo findDebugInfoFile(const QString &root, const QString &file,
...
+        if (debugLinkFile.isFile()) {
             return debugLinkFile;
+        }
     }
 
-    debugLinkFile.setFile(dir, file + QDir::separator() + QLatin1String(".debug")
-                                + QDir::separator() + debugLinkString);
-    if (debugLinkFile.isFile())
+    debugLinkFile.setFile(dir.path() + QDir::separator() + QLatin1String("usr") + QDir::separator()
+                          + QLatin1String("lib") + QDir::separator() + QLatin1String("debug") + QDir::separator()
```
Comment (dfaure-kdab) [inline app/perfsymboltable.cpp:507]: Why all the QDir::separator() business? Not only those paths are very Unix specific, but anyway QFile supports '/' on Windows too.

File: app/perfsymboltable.cpp [hunk 16]
```diff
@@ -508,19 +536,23 @@ int PerfSymbolTable::findDebugInfo(Dwfl_Module *module, const char *moduleName,
...
     // fall-back, mostly for situations where we loaded a file via it's build-id.
     // search all known paths for the debug link in that case
-    const auto debugLinkString = QFile(debugLink).fileName();
+    // const auto debugLinkString = QFile(debugLink).fileName();
     auto debugLinkFile = findFile(debugLink, debugLinkString);
     if (!debugLinkFile.isFile()) {
         // fall-back to original file path with debug link file name
-        const auto &elf = m_elfs.findElf(base);
-        const auto &path = QString::fromUtf8(elf.originalPath);
+        // const auto &elf = m_elfs.findElf(base);
```
Comment (milianw) [inline app/perfsymboltable.cpp]: dito

**Review by milianw — CHANGES_REQUESTED** [review]

File: app/perfsymboltable.cpp (hunk 15 above)
Comment (lievenhey) [reply app/perfsymboltable.cpp:507]: I wanted to keep the file in the same style

Comment (milianw) [reply app/perfsymboltable.cpp:507]: @dfaure-kdab I'm with you here, but this was TQC Ulf's choice back then when he ported this to windows - it's pretty mysterious to me too why this is the way it is. I never tested it on windows myself though...

**Comment by milianw** [comment]
I'd say let's go for it :) thanks!
